coqlspclient/proof_file.py

```diff
@@ -410,15 +471,81 @@ def proofs(self) -> List[ProofTerm]:
         """
         return self.__proofs
 
-    @property
-    def context(self) -> FileContext:
-        """
-        Returns:
-            FileContext: Whole context available in the environment of the Coq file.
-        """
-        return self.coq_file.context
+    def add_step(self, step_text: str, previous_step_index: int):
+        proof, prev = self.__find_prev(self.steps[previous_step_index].ast.range)
+        if prev == -1:
+            self._make_change(self._add_step, step_text, previous_step_index)
+        else:
+            self._make_change(self._add_step, step_text, previous_step_index, True)
+        proof.steps.insert(prev + 1, self.__get_step(previous_step_index + 1))
+        # We should change the goals of all the steps in the same proof
+        # after the one that was changed
+        # NOTE: We assume the proofs and steps are in the order they are written
+        for e in range(prev + 2, len(proof.steps)):
+            # The goals will be loaded if used (Lazy Loading)
+            proof.steps[e].goals = self._goals
+
+    def delete_step(self, step_index: int) -> None:
+        step = self.steps[step_index]
+        optional = self.__find_step(step.ast.range)
+        if optional is not None:
+            proof, i = optional
+            proof.steps.pop(i)
+            for e in range(i, len(proof.steps)):
+                # The goals will be loaded if used (Lazy Loading)
+                proof.steps[e].goals = self._goals
+        else:
+            raise NotImplementedError(
+                "Deleting steps outside of a proof is not implemented yet"
+            )
+        self._make_change(self._delete_step, step_index, True)
+
+    def change_steps(self, changes: List[CoqChange]):
+        offset_steps = 0
+        previous_steps_size = len(self.steps)
+
+        for change in changes:
+            if isinstance(change, CoqAddStep):
+                proof, i = self.__find_prev(
+                    self.steps[change.previous_step_index].ast.range
+                )
+                self._add_step(
+                    change.step_text,
+                    change.previous_step_index,
+                    in_proof=i is not None,
```

Comment:
`__find_prev` no longer returns a None index. 
```suggestion
                    in_proof=i > -1,
```

Reply:
Actually with the new `__find_prev` this should always be true. Because now we are sure that either the  `__find_prev` raises an error or the step to add will always be inside a proof.

Reply:
```suggestion
                    in_proof=True,
```